intro/blockchain-intro-solution.js

```diff
@@ -0,0 +1,69 @@
+let SHA512 = require('crypto-js/sha512');
+
+class Block {
+  constructor(index, timestamp, data, previousHash = '') {
+    this.index = index;
+    this.timestamp = timestamp;
+    this.data = data;
+    this.previousHash = previousHash;
+    this.hash = this.calculateHash();
+    this._nonce = 0;
+  }
+
+  // Calculate a hash based on data supplied to block
+  calculateHash() {
+    return SHA512(`${this.index}${this.previousHash}${this.timestamp}${JSON.stringify(this.data)}${this._nonce}`).toString();
+  }
+
+  // Mine block by difficulty
+  mineBlock(difficulty) {
+    console.log('Mining block...');
+    while (this.hash.substring(0, difficulty) !== Array(difficulty + 1).join('0')) {
+      this._nonce++;
+      this.hash = this.calculateHash();
+    }
+    console.log(`${'Block mined'} ${this.hash}`);
+  }
+}
+
+class Blockchain {
+  constructor() {
+    this.chain = [this._createGenesisBlock()];
+    this._difficulty = 3;
+  }
+
+  // Create the initial block for the blockchain
+  _createGenesisBlock() {
+    return new Block(0, (new Date()).toString(), 'Genesis Block', '0');
+  }
+
+  // Get the last created block
+  getLatestBlock() {
+    return this.chain[this.chain.length - 1];
+  }
+
+  // Add a new block to the blockchain
+  addBlock(newBlock) {
+    newBlock.previousHash = this.getLatestBlock().hash;
+    newBlock.mineBlock(this._difficulty);
+    this.chain.push(newBlock);
+  }
+
+  // Check if the chain is valid
+  isValidChain() {
+    let validity = true;
+
+    this.chain.forEach((block, i) => {
```

Comment:
this might be a fun opportunity to use `Array.prototype.every`!
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every

eg, `return this.chain.every(isValidBlock);` (where `isValidBlock` would be an inline function or a separate helper function)

Reply:
Hooray for new native Array methods!

Reply:
Oh nice. I wasn't even aware of this method

Reply:
Updated